### PR TITLE
client: respect context deadline in attach

### LINKED ISSUE #36543: client: container attach dial does not timeout
no logs to show, I do have a test case. a PR is incoming.

the issue: container attach does not timeout the dial or the initial http request before hijacking, not respecting the context.Context's deadline that it has received as a parameter.

### PR DESCRIPTION
**Pull request by rdallman**
ContainerExecAttach and ContainerAttach both use postHijacked to hijack a
connection and do hijack things. However, postHijacked, while taking a
context.Context as a param, its callees do not obey this context, of note:
dial and the initial http request before hijacking. This is an attempt to
remedy that. Unfortunately, most of the underlying methods themselves do not
accept a context and this was not simply a matter of passing that context
along -- this isn't very pretty and I'm open to ideas on how best to do this.

there is a test that verifies this behavior from dial now. I hope the ci tests are good enough to catch the existing happy paths :)

thank you all for your work!

fixes #36543 

timeout_spirit_animal

### CODE REVIEW

**Comment by rdallman** [comment]
i see the tests need a little platform detection, will try to resolve tomorrow (with anything else that explodes :)

File: client/hijack.go [hunk 1]
```diff
@@ -140,23 +167,37 @@ func tlsDialWithDialer(dialer *net.Dialer, network, addr string, config *tls.Con
...
+
 	if tlsConfig != nil && proto != "unix" && proto != "npipe" {
 		// Notice this isn't Go standard's tls.Dial function
-		return tlsDial(proto, addr, tlsConfig)
+		return tlsDialWithDialer(&net.Dialer{Timeout: timeout}, proto, addr, tlsConfig)
 	}
 	if proto == "npipe" {
-		return sockets.DialPipe(addr, 32*time.Second)
+		if timeout == 0 {
+			// Why 32? See issue 8035
```
Comment (thaJeztah) [inline client/hijack.go]: Oh, LOL, I recall that discussion :joy: https://github.com/moby/moby/pull/8035#issuecomment-55645549

**Comment by rdallman** [comment]
@thaJeztah alright I've been kicking this the last couple of days w/ rebasing tip and it seems the tests waffle back and forth on working on some seemingly unrelated tests, so I think this is OK for review now (the history of master builds seems to confirm this). commits are all signed as well to appease the shelled one.

PTAL, thank you!

**Comment by thaJeztah** [comment]
ping @tonistiigi @cpuguy83 PTAL

File: client/hijack.go [hunk 2]
```diff
@@ -140,23 +167,37 @@ func tlsDialWithDialer(dialer *net.Dialer, network, addr string, config *tls.Con
 	return &tlsClientCon{conn, rawConn}, nil
 }
 
-func dial(proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+func dial(ctx context.Context, proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+	// 0 = no timeout
+	var timeout time.Duration
+
+	deadline, ok := ctx.Deadline()
```
Comment (tonistiigi) [inline client/hijack.go]: Can we leave this only for the `npipe` case and everything else can use the context directly?

File: integration-cli/docker_cli_run_test.go [hunk 3]
```diff
@@ -4343,6 +4343,9 @@ func (s *DockerSuite) TestRunEmptyEnv(c *check.C) {
 
 // #28658
 func (s *DockerSuite) TestSlowStdinClosing(c *check.C) {
+	// TODO Windows: This needs some work to make compatible.
+	testRequires(c, DaemonIsLinux)
```
Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: Why is this change here? Does this PR change anything about the http API?

File: client/hijack.go [hunk 4]
```diff
@@ -175,15 +216,30 @@ func (cli *Client) setupHijackConn(req *http.Request, proto string) (net.Conn, e
 	defer clientconn.Close()
 
 	// Server hijacks the connection, error 'connection closed' expected
-	resp, err := clientconn.Do(req)
-	if err != httputil.ErrPersistEOF {
+	errCh := make(chan error, 1)
+	go func() {
+		resp, err := clientconn.Do(req)
```
Comment (tonistiigi) [inline client/hijack.go]: I'm not sure I understand this. Why doesn't `req.WithContext` already handle this?

Comment (rdallman) [reply client/hijack.go]: sadly, the `*httputil.ClientConn` methods do not respect the request context, per https://golang.org/src/net/http/httputil/persist.go?s=11092:11159#L415 -- there is a loud note on its factory method about how it should not be part of stdlib and a user should use methods in the `net/http` package and this library like won't be updated, it seems. I briefly investigated the possibility of using `net/http`, but there's no clear path using `net/http` methods to do what we want here.

File: integration-cli/docker_cli_run_test.go (hunk 3 above)
Comment (rdallman) [reply integration-cli/docker_cli_run_test.go]: this could be just tests waffling, I was hoping to get potentially more insight from someone familiar with windows happenings here. maybe that answers my question, though.

File: client/hijack.go (hunk 2 above)
Comment (rdallman) [reply client/hijack.go]: `tlsDialWithDialer` requires translating to a timeout, as well. `net.DialContext` is the 1 exception here, unfortunately. I'm happy to:

*  move context->timeout translation down to `tlsDialWithDialer` & the `npipe` block from here
*  change `tlsDialWithDialer` to `tlsDialWithContext`

sounds good?

Comment (tonistiigi) [reply client/hijack.go]: `tlsDialWithContext` is what I thought. If we keep using the dialer, possibly the correct solution would be to wrap it early with context cancellation calling `dialer.Cancel` instead of the timeout logic.

Generally, if the stdlib or helper packages do not provide good ways to implement some cases we should try to fix them in these packages and ignore these cases until then. Having these exceptions and potentially leaking goroutines is worse than a little delay on cancellation. For example, I don't think the extra goroutine in `Handshake` is justified. My suggestion would be to either use the context directly and ignore the cases that can't do `DialContext` or convert to dialer once and only use dialer for everything.

File: integration-cli/docker_cli_run_test.go (hunk 3 above)
Comment (tonistiigi) [reply integration-cli/docker_cli_run_test.go]: I think the windows flakiness for this test was fixed with #36571

**Comment by kolyshkin** [comment]
CI failure on experimental ([logs](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/39779/console)) is unrelated; please see https://github.com/moby/moby/issues/36547.

File: client/container_exec_test.go [hunk 5]
```diff
@@ -79,6 +81,28 @@ func TestContainerExecStartError(t *testing.T) {
...
+		}),
+
+		proto:    hostURL.Scheme,
+		addr:     hostURL.Host,
+		basePath: hostURL.Path,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Nanosecond)
+	defer cancel()
+	_, err = client.ContainerExecAttach(ctx, "nothing", types.ExecStartCheck{})
+	if err == nil || (!strings.Contains(err.Error(), "i/o timeout") && // unix
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply client/container_exec_test.go]: Something like `if !errors.Cause(err).(net.Error).Timeout() { }` seems like it'd work.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: client/hijack.go (hunk 2 above)
Comment (rdallman) [reply client/hijack.go]: I did some research on this, and pursuant to https://github.com/rdallman/moby/blob/ccd30e464ac89d8bb2b4b64fb7b57128c721dc52/client/hijack.go#L58-L62 I believe that we can use https://golang.org/pkg/crypto/tls/#Conn.CloseWrite now in stdlib, and toss out the custom implementation here to be able to close the write part of the connection. woo. I will try to go ahead with this, and move this code to a `client/hijack_go17.go` file for older go versions as seems the convention here https://github.com/rdallman/moby/blob/ccd30e464ac89d8bb2b4b64fb7b57128c721dc52/client/tlsconfig_clone_go17.go -- we will still convert our context deadline into a dial deadline, but this seems fine in the grand scheme of things.

the code itself does spawn a goroutine https://golang.org/src/crypto/tls/tls.go?s=3255:3347#L148 the same as we do for the handshake timeout.

this leaves us with only the hijack do goroutine explicitly in the client code. after some auditing, it appears to use `(http.Request).Write` and `http.ReadResponse` which will respect the context, and I will remove that addition, thanks for pointing this out.

File: client/container_exec_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdallman) [reply client/container_exec_test.go]: works great on linux, thanks. windows:

```
21:15:12 --- FAIL: TestContainerExecAttachTimeout (0.00s)
21:15:12 	container_exec_test.go:103: expected a timeout error, got cannot connect to the Docker daemon. Is 'docker daemon' running on this host?: dial tcp 0.0.0.0:4243: socket: The requested service provider could not be loaded or initialized.
```

will attempt to find something suitable, feedback loops on win are long for me, sorry. open to ideas meanwhile.

[the remark this answers is not in the document]
Comment (thaJeztah) [reply client/container_exec_test.go]: ping @johnstep ptal ^^

**Comment by mlaventure** [comment]
ping @rdallman @cpuguy83 @tonistiigi 

What's happening there? Looks like this should be ready to merge

**Comment by rdallman** [comment]
thanks @mlaventure i believe there are some hangups https://github.com/moby/moby/pull/36544#discussion_r174307775 -- i've begun investigating skirting around the `httputil.ClientConn` so that we can use the `net/http` clients for tls dialing, they handle socks5 now in stdlib and we will then no longer need (i think) https://github.com/docker/go-connections which makes us have to add some workarounds to plumb the context through -- in any case, this seems doable with some `http.RoundTripper` magic, but I've not got too much to show for it at the moment. without fixing the go-connections timeout/context plumbing issue this remains incomplete, and there is a (good) aversion to launching goroutines to handle timeouts ourselves in the client.

**Comment by rdallman** [comment]
I may one day get back to this, but I feel bad leaving this laying around now. the stack has grown deep, friends. thanks all for review/project, maybe someone else gets around to finishing this up :)
